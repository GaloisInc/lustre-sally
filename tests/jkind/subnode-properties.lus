-- This file has one top level property which requires subnode
-- properties in order to prove. It's somewhat obfuscated to avoid
-- automatic invariant generation.

node counter(rand : bool) returns (x : int);
var
  lemma : bool;
let
  x = (if rand then -1 else -1) -> 1 + pre x;

  lemma = x >= -1;
  --%PROPERTY lemma;
tel;

node counter2(rand : bool) returns (x : int);
var
  lemma : bool;
let
  x = (if rand then -1 else -1) -> 1 + pre x;

  lemma = x >= -1;
  --%PROPERTY lemma;
tel;

node ten(rand : bool) returns (ten : bool);
let
  ten = counter(rand) >= 10;
tel;

node main(rand : bool) returns (prop : bool);
let
  prop = counter2(rand) < 10 or ten(rand);
  --%PROPERTY prop;
tel;
