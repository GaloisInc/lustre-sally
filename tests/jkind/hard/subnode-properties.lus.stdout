Core Lustre
===========

node (rand_12 : bool)
  returns (prop)
  assumes ()
  shows (prop, lemma_23, lemma_20)
let
  rand_19 = rand_12
  valof_pre___no_static_25 = pre __no_static_15
  __no_static_15 = ITE(rand_19, Neg(1), Neg(1)) -> Add(1, valof_pre___no_static_25)
  rand_21 = rand_12
  rand_22 = rand_21
  valof_pre___no_static_26 = pre __no_static_24
  __no_static_24 = ITE(rand_22, Neg(1), Neg(1)) -> Add(1, valof_pre___no_static_26)
  __no_static_16 = Geq(__no_static_24, 10)
  prop = Or(Lt(__no_static_15, 10), __no_static_16)
  lemma_20 = Geq(__no_static_15, Neg(1))
  lemma_23 = Geq(__no_static_24, Neg(1))
tel
Sally Model
===========

(define-state-type
   S
   (
     (prop Bool)
     (|__no_static:15:init| Bool)
     (|__no_static:15| Int)
     (|__no_static:16| Bool)
     (|__no_static:24:init| Bool)
     (|__no_static:24| Int)
     (|gal-initializing| Bool)
     (|lemma:20| Bool)
     (|lemma:23| Bool)
     (|rand:12| Bool)
     (|rand:19| Bool)
     (|rand:21| Bool)
     (|rand:22| Bool)
     (|valof_pre___no_static:25| Int)
     (|valof_pre___no_static:26| Int))
   (
     (|rand:12| Bool)))
(define-transition-system
   TS
   S
   (and
      (= |gal-initializing| true)
      (= |__no_static:15:init| true)
      (= |__no_static:24:init| true))
   (and
      (= next.|gal-initializing| false)
      (= next.|rand:12| input.|rand:12|)
      (= next.|rand:19| next.|rand:12|)
      (= next.|valof_pre___no_static:25| state.|__no_static:15|)
      (ite
         state.|__no_static:15:init|
         (=
            next.|__no_static:15|
            (ite
               next.|rand:19|
               (- 1)
               (- 1)))
         (=
            next.|__no_static:15|
            (+ 1 next.|valof_pre___no_static:25|)))
      (= next.|__no_static:15:init| false)
      (= next.|rand:21| next.|rand:12|)
      (= next.|rand:22| next.|rand:21|)
      (= next.|valof_pre___no_static:26| state.|__no_static:24|)
      (ite
         state.|__no_static:24:init|
         (=
            next.|__no_static:24|
            (ite
               next.|rand:22|
               (- 1)
               (- 1)))
         (=
            next.|__no_static:24|
            (+ 1 next.|valof_pre___no_static:26|)))
      (= next.|__no_static:24:init| false)
      (=
         next.|__no_static:16|
         (<= 10 next.|__no_static:24|))
      (=
         next.prop
         (or
            (< next.|__no_static:15| 10)
            next.|__no_static:16|))
      (=
         next.|lemma:20|
         (<=
            (- 1)
            next.|__no_static:15|))
      (=
         next.|lemma:23|
         (<=
            (- 1)
            next.|__no_static:24|))))

(query
   TS
   (or |gal-initializing| prop))
(query
   TS
   (or |gal-initializing| |lemma:23|))
(query
   TS
   (or |gal-initializing| |lemma:20|))

[Lustre] Validating properties:
[Lustre] Property prop... considering simultaneous states to depth 0
considering simultaneous states to depth 0
considering simultaneous states to depth 1
considering simultaneous states to depth 1
considering simultaneous states to depth 2
considering simultaneous states to depth 2
considering simultaneous states to depth 3
considering simultaneous states to depth 3
considering simultaneous states to depth 4
considering simultaneous states to depth 4
considering simultaneous states to depth 5
considering simultaneous states to depth 5
considering simultaneous states to depth 6
considering simultaneous states to depth 6
considering simultaneous states to depth 7
considering simultaneous states to depth 7
considering simultaneous states to depth 8
considering simultaneous states to depth 8
considering simultaneous states to depth 9
considering simultaneous states to depth 9
considering simultaneous states to depth 10
considering simultaneous states to depth 10
counter-example search depth 0
counter-example search depth 1
counter-example search depth 2
counter-example search depth 3
counter-example search depth 4
counter-example search depth 5
counter-example search depth 6
counter-example search depth 7
counter-example search depth 8
counter-example search depth 9
counter-example search depth 10
[[1;93mUnknown[0m] Valid up to depth 10
[Lustre] Property lemma... considering simultaneous states to depth 0
considering simultaneous states to depth 0
considering simultaneous states to depth 1
considering simultaneous states to depth 1
[[1;92mValid[0m] considering simultaneous states to depth 1
[Lustre] Property lemma... considering simultaneous states to depth 0
considering simultaneous states to depth 0
considering simultaneous states to depth 1
considering simultaneous states to depth 1
[[1;92mValid[0m] considering simultaneous states to depth 1
[Lustre] Model status: [[1;93mUnknown[0m] 
